cpp/src/arrow/util/thread-pool.h

```diff
@@ -110,30 +93,18 @@ class ARROW_EXPORT ThreadPool {
     return SpawnReal(std::forward<Function>(func));
   }
 
-  // Submit a callable and arguments for execution.  Return a future that
-  // will return the callable's result value once.
-  // The callable's arguments are copied before execution.
-  // Since the function is variadic and needs to return a result (the future),
-  // an exception is raised if the task fails spawning (which currently
-  // only occurs if the ThreadPool is shutting down).
-  template <typename Function, typename... Args,
-            typename Result = typename std::result_of<Function && (Args && ...)>::type>
-  std::future<Result> Submit(Function&& func, Args&&... args) {
-    // Trying to templatize std::packaged_task with Function doesn't seem
-    // to work, so go through std::bind to simplify the packaged signature
-    using PackagedTask = std::packaged_task<Result()>;
-    auto task = PackagedTask(std::bind(std::forward<Function>(func), args...));
-    auto fut = task.get_future();
-
-    Status st = SpawnReal(detail::packaged_task_wrapper<Result>(std::move(task)));
+  // Submit a callable for execution.
+  void Submit(std::function<Status()> func) {
+    Status st = SpawnReal(std::move(func));
     if (!st.ok()) {
       // This happens when Submit() is called after Shutdown()
       std::cerr << st.ToString() << std::endl;
       std::abort();
     }
-    return fut;
   }
 
+  void Wait();
```

Comment:
I'd also expose the WaitFor variant with chrono (see https://en.cppreference.com/w/cpp/thread/condition_variable/wait_for).